chore(team): write governance (#159)
nodejs/userland-migrations: TEAM.md

> # Team membership

Reviewer: ```suggestion
# Governance
```

Reviewer: This file should be named `GOVERNANCE.md`

Author: is it really a governance scince we are a team ?

Reviewer: Yes. This file documents the governance of this team and repository, and hence, it should be in a governance file

Current members are:

> - `@alexbit-codemod`

Reviewer 2: please add @mohebifar to the list here. you might need to send him a new invite for the node org since we missed the initial deadline. sry about that. thanks team.

Reviewer 3: omg 3rd time? 🤦 sorry about that

Reviewer 3: 🎉🎉 Thank you! Joined.

> - `@alexbit-codemod`

Reviewer 4: Yes. IIR whilst we were getting Mo sorted.

(I haven't removed him from the GH team)

- `@AugustinMauroy`
- `@avivkeller`
- `@bmuenzenmeyer`

> - `@JakobJingleheimer` (lead)

Reviewer: ```suggestion
- `@JakobJingleheimer` (Maintainer)
```

Also, let's follow the format of other repositories, that is:
```md
- [@avivkeller](https://github.com/avivkeller) - **Aviv Keller** (he/him)
```

- `@ljharb`

## Rights

> Any team member may block a PR or decision. A block may be overturned by team vote; votes require a quorum of ⅔ of active members and pass by >½ simple majority. In accordance with Node.js policy, a team's vote may be overturned by the [TSC](https://github.com/nodejs/TSC).

Reviewer: ```suggestion
Any team member can block a pull request or decision. A block can be overturned by a team vote, which requires a quorum of two-thirds of active members and passes with a simple majority (>50%). Per Node.js policy, the [TSC](https://github.com/nodejs/TSC) may override a team’s vote.
```

Reviewer 4: Nit: `Per Node.js policy` → `Pursuant to Node.js policy`

> Any team member may block a PR or decision. A block may be overturned by team vote; votes require a quorum of ⅔ of active members and passes by >½ simple majority. In accordance with Node.js policy, a team's vote may be overturned by the [TSC](https://github.com/nodejs/TSC).

Reviewer 4: ```suggestion
Any team member may block a PR or decision. A block may be overturned by team vote; votes require a quorum of ⅔ of active members and pass by >½ simple majority. In accordance with Node.js policy, a team's vote may be overturned by the [TSC](https://github.com/nodejs/TSC).
```

> Any member can create a tag on main which triggers a CI run, this is used to create a releases. So be careful with this power. Tags should have this structure `vX.Y.Z@workspace`, where `X.Y.Z` is the version number and `workspace` is the name of the npm workspace (e.g. `v1.2.3@codemod`).

Reviewer: ```suggestion
Any member can create a tag on the repository's `HEAD` branch, which triggers a CI run and initiates a release. Tags must follow the format `vX.Y.Z@workspace`, where `X.Y.Z` is the version number and `workspace` is the name of the npm workspace (e.g., `v1.2.3@codemod`).
```

Reviewer 4: No `,` after `e.g.` though

> Any member can create a tag on main which triggers a CI run, this is used to create a releases. So be careful with this power. Tags should have this structure `vX.Y.Z@workspace`, where `X.Y.Z` is the version number and `workspace` is the name of the npm workspace (e.g. `v1.2.3@codemod`).

Author: okay let's give release thing to lead guys !

> Any member can create a tag on main which triggers a CI run, this is used to create a releases. So be careful with this power. Tags should have this structure `vX.Y.Z@workspace`, where `X.Y.Z` is the version number and `workspace` is the name of the npm workspace (e.g. `v1.2.3@codemod`).

Reviewer 4: Let's add another section here for who can be a releaser. I'm thinking maybe the only extra criteria needed is collaborator.

> ### Active vs inactive membership
> 
> An active member is one who has substantively participated once in the past 6 weeks.
> 
> A member who has been inactive for 4 or more months may be removed. They are welcome to return by simple request.

Reviewer: ```suggestion
### Retaining Membership

Members are expected to abide by all guidelines set in place by the Node.js Project, including, but not limited to:
- The [Node.js Moderation Policy](https://github.com/nodejs/admin/blob/HEAD/Moderation-Policy.md)
- The [Code of Conduct](https://github.com/nodejs/node/blob/HEAD/CODE_OF_CONDUCT.md)

Violations of the Node.js Project's guidelines may result in immediate removal depending on severity and TSC decision.

Members are expected to engage in project discussions regularly, and members inactive for more than six months may be removed from the active members list. Removed members an be reinstated upon returning to active participation.
```

Reviewer 4: > Violations of the Node.js Project's guidelines may result in immediate removal depending on severity and TSC decision.

This repeats things from the Node.js policy. I think it should be left there.

`Removed members an be` → `Removed members can be`

> inactive for more than six months

I was thinking to align with Node.js's policy for active collaborators (which is 4 months).

Reviewer: > six months

I was taking the language from the web-team information, the "six" can be whatever we want

> I was thinking to align with Node.js's policy for active collaborators (which is 4 months).

I thought it was 9?

Reviewer 4: It's definitely not 9; I think it was 6 but is 4 now

#### Team lead

The team lead is not more equal than other active members; it is an administrative function. For administrative purposes, the lead owns the repo, the github team, and handles membership changes and repository administration (particularly potentially destructive actions).

The team lead is responsible for ensuring team policies are followed, and takes some discretion when taking administrative action (such as removing an inactive member per team policy).

> The team lead can also bypass Github Branch rulesets to merge PRs that are otherwise blocked by the rules. Only for hotfixes.

Reviewer 4: ```suggestion
The team lead can also bypass Github Branch rulesets to merge PRs that are otherwise blocked by the rules (such as for trivial edits/corrects and hot-fixes).
```

> The team lead can also bypass Github Branch rulesets to merge PRs that are otherwise blocked by the rules (such as for trivial edits/corrects and hot-fixes).

Reviewer: ```suggestion
#### Team Maintainers

Team maintainers serve in an administrative capacity and are not considered "more equal" than other active contributors. Their primary role is to support smooth project operations.

Maintainership includes responsibilities such as managing the GitHub repository and team, handling membership changes, and performing administrative tasks, especially those that may be destructive or sensitive in nature.

Maintainers ensure team policies are upheld and may exercise discretion when taking necessary actions, such as removing inactive members in accordance with team guidelines.

Additionally, maintainers have the ability to bypass GitHub branch protection rules to merge pull requests, typically in cases of trivial edits, urgent corrections, or hot-fixes.
```

Reviewer: Using the term "team lead" implies that one person is in charge (or, well, a leader). In our case, while @JakobJingleheimer _is_ the "lead", it implies that another person wouldn't be able to _also_ be a maintainer.

## Team nomination

This policy extends the parent Node.js policy.

Current Node.js collaborators are free to join at any time and do not require a nomination.

Current Node.js members who are not collaborators are free to request to join without a nomination or meeting contribution criteria. The request must have no objections after 3 days. All active members must be notified of the request.

Non-members of the Node.js org should meet the following criteria:

- ~2 months of consistent (substantive, non-trivial) contribution/participation.
  - _Consistent_ means roughly once a week (we all have lives)
  - _Substantive_ means adds value and does not detract from the issue at hand.

> Exceptions may be made to members of the larger ecosystem who are well known to the Node.js organisation.

Reviewer 4: ```suggestion
Exceptions may be made for members of the larger ecosystem who are well known to the Node.js organisation.
```

An active member may nominate a contributor who meets this criteria. The nomination must pass before requesting the contributor be added to the organisation. All active members must be notified of the nomination.

> A nomination should be raised privately with active members before publicly, out of respect for the nominee.

Reviewer: I disagree with this:

1. Nomination should be required for all members, or a members list should not be kept. (I'm okay with either)
2. We shouldn't be setting guidelines for what it means to be nominated. Users should contribute, and be nominated if the team sees fit, not based on criteria

Wording is _very_ important here as well.

Author: cc @JakobJingleheimer that your word

Reviewer 4: > Nomination should be required for all members

Node.js policy says collaborators may join any team. We can't just override that. I think we can do the "freeze" thing.

The exceptions for existing non-collaborator members and for ecosystem members well known to the org is for instance how we have you and Mo; it would be kind of silly to say "sorry author of the tool we use, we need to see you consistently demonstrating expertise". It would be up to the team's discretion to grant an exception.

> or a members list should not be kept

Hmm? Why? If a collaborator wants to join the team, they add themselves to the list (or a maintainer does it as part of adding them to the GH team). Or have I misunderstood you?

> We shouldn't be setting guidelines for what it means to be nominated. Users should contribute, and be nominated if the team sees fit, not based on criteria

We've suffered a lack of criteria for collaborator nomination for years, and it's been especially problematic recently (it's in the process of being addressed now). It creates a huge amount of animosity whenever there's a disagreement/objection. I think it would be a big mistake to repeat that same problem here.

I strongly think we need some kind of baseline; it doesn't have to be what I had jotted down though (I said it was a starting point / draft). In addition to avoiding the aforementioned animosity, it also shows hopeful contributors where the bar/goalpost is (nobody appreciates vague goalposts).

Reviewer: > Or have I misunderstood you?

My main concern is that a member will join the team without adding themselves to the list, but perhaps that's a silly worry.

> I think it would be a big mistake to repeat that same problem here.

I see. I feel like the current wording is way too rigid, i.e. "_Consistent_ means roughly once a week (we all have lives)" is still, IMO, too often.

Something _slightly_ vaguer, but still firm may help:
```
Individuals outside the organization (or however we define non-members) who are interested in joining the team are encouraged to demonstrate consistent and meaningful participation.

Consistent meaning regular engagement over time, not necessarily every week, but enough to show sustained interest and involvement.

Meaningful meaning contributions that add clear value and positively support the work being done.
```

> ## Team Expulsion
> 
> A team member who violates Node.js's code of conduct or who is acting against the interests or mandate of the team, or acts in bad faith may be expelled by vote of the team; such a vote requires a quorum of ⅔ of active members passing by ⅗ super majority. If a vote passes a simple majority but fails a super majority, the team member will be asked to voluntarily exit. Expulsion votes are confidential and are not conveyed to the expulee (they may be verified in confidence by the TSC).
> 
> When expulsion proceedings are commenced, team membership is frozen until the vote has settled.

Reviewer: ```suggestion
```

See https://github.com/nodejs/userland-migrations/pull/159/files#r2258269636

Reviewer 4: This is not quite the same, and if someone turns out to be toxic to the team, we need an explicit way to remove them. The TSC is the escalation when things can't be sorted, not the first and only lever. Similarly, the moderation team handles things for the org at large. A CoC violation may be minor in terms of "this person isn't banned from Node.js", but that doesn't mean the team has to keep them.

Teams have failed to set this up in the past, and it killed those teams (basically everyone quit). Hope for the best, but plan for the worst.
